### ISSUE TITLE
Selection rectangle not appearing

### ISSUE THREAD

**Issue by hamzawain7** [fabricjs/fabric.js#6954]
## Version
4.3.1

## Test Case
https://jsfiddle.net/hamzawain7/45d2sr3y/

## Information about environment
Nodejs or browser?
chrome

## Steps to reproduce
Try to create a selection box by click and drag. It doesn't appear because the initAligningGuidelines library keeps on clearing contextTop

## Expected Behavior
Group selection rectangle should be visible

## Actual Behavior
Group selection rectangle isn't be visible when using the aligning_guidelines.js or centering_guidelines.js libraries

**Comment by asturur** [comment]
The use of contextTop is pretty much unstable. in the roadmap there was the idea of refactoring it to be a pipeline so that devs could queue functions for it, and having fabriJS just handling the lifecycle of the render.

I won't have time to work on this soon, if someone wants to try to fix i m available to help.

**Comment by graphics-et-al** [comment]
I'm having some real issues with this too. Is it on the roadmap?

**Comment by graphics-et-al** [comment]
Upon further examination, the issue seems to be with the "canvas.renderAll();" function.
If you change  to:

fabric.util.requestAnimFrame(function render() {
        //canvas.renderAll();
        fabric.util.requestAnimFrame(render);
    });

the selection rectangle works as expected. But, then the video won't play *unless you're moving it*

So, the issue seems to be the selection rectangle gets cleared when canvas.renderAll(); or canvas.requestRenderAll(); gets called.

**Comment by asturur** [comment]
Sorry for not coming back here.
So yes, renderAll will clearContextTop.
The selection uses requestRenderAll() that will schedule a render at the next frame, and the selection is state based, meaning that when you move the mouse the selection area is described in an object that then gets rendered.

Did you try by chance using `canvas.requestRenderAll in the animation callback?

**Comment by ShaMan123** [comment]
related #6873

**Comment by graphics-et-al** [comment]
> Sorry for not coming back here. So yes, renderAll will clearContextTop. The selection uses requestRenderAll() that will schedule a render at the next frame, and the selection is state based, meaning that when you move the mouse the selection area is described in an object that then gets rendered.
> 
> Did you try by chance using `canvas.requestRenderAll in the animation callback?

No- but I will give it a go! As a workaround I put mouse down, out and up handlers that toggle a little variable:

`
    canvas.on('mouse:down', (options) => {
        isMouseDown = true;
    });

    canvas.on('mouse:up', (options) => {
        isMouseDown = false
    });

    canvas.on('mouse:out', (options) => {
        isMouseDown = false
    });
`
so, we can not do the offending action when the mouse is down

`fabric.util.requestAnimFrame(function render() {
                canvas.requestRenderAll();
                fabric.util.requestAnimFrame(render);
          }, fab_video);`

This works for now, but it does pause the video when selecting/moving etc. Which isn't actually all that bad...

**Comment by w934423231** [comment]
Is there any good solution to the current problem? I have also encountered this problem

**Comment by ShaMan123** [comment]
https://github.com/fabricjs/fabric.js/blob/master/lib/aligning_guidelines.js#L193

This line is causing the issue

Replaced it with a small check
https://jsfiddle.net/7nb3qf5t/
